src/manager/runner_manager.py

```diff
@@ -0,0 +1,275 @@
+from metrics import github as github_metrics
+from metrics import runner as runner_metrics
+from metrics.runner import RunnerMetrics
+
+logger = logging.getLogger(__name__)
+
+IssuedMetricEventsStats = dict[Type[metric_events.Event], int]
+
+
+class FlushMode(Enum):
+    """Strategy for flushing runners.
+
+    Attributes:
+        FLUSH_IDLE: Flush idle runners.
+        FLUSH_BUSY: Flush busy runners.
+    """
+
+    FLUSH_IDLE = auto()
+    FLUSH_BUSY = auto()
+
+
+@dataclass
+class RunnerInstance:
+    """Represents an instance of runner.
+
+    Attributes:
+        name: Full name of the runner. Managed by the cloud runner manager.
+        id: ID of the runner. Managed by the runner manager.
+        github_state: State on github.
+        cloud_state: State on cloud.
+    """
+
+    name: str
+    id: InstanceId
+    github_state: GithubRunnerState | None
+    cloud_state: CloudRunnerState
+
+    def __init__(self, cloud_instance: CloudRunnerInstance, github_info: SelfHostedRunner | None):
+        """Construct an instance.
+
+        Args:
+            cloud_instance: Information on the cloud instance.
+            github_info: Information on the GitHub of the runner.
+        """
+        self.name = cloud_instance.name
+        self.id = cloud_instance.id
+        self.github_state = (
+            GithubRunnerState.from_runner(github_info) if github_info is not None else None
+        )
+        self.cloud_state = cloud_instance.state
+
+
+@dataclass
+class RunnerManagerConfig:
+    """Configuration for the runner manager.
+
+    Attributes:
+        token: GitHub personal access token to query GitHub API.
+        path: Path to GitHub repository or organization to registry the runners.
+    """
+
+    token: str
+    path: GithubPath
+
+
+class RunnerManager:
+    """Manage the runners."""
+
+    def __init__(self, cloud_runner_manager: CloudRunnerManager, config: RunnerManagerConfig):
+        """Construct the object.
+
+        Args:
+            cloud_runner_manager: For managing the cloud instance of the runner.
+            config: Configuration of this class.
+        """
+        self._config = config
+        self._cloud = cloud_runner_manager
+        self._github = GithubRunnerManager(
+            prefix=self._cloud.get_name_prefix(), token=self._config.token, path=self._config.path
+        )
+
+    def create_runners(self, num: int) -> tuple[InstanceId]:
+        """Create runners.
+
+        Args:
+            num: Number of runners to create.
+
+        Returns:
+            List of instance ID of the runners.
+        """
+        logger.info("Creating %s runners", num)
+        registration_token = self._github.get_registration_token()
+
+        create_runner_args = [RunnerManager._CreateRunnerArgs(self._cloud, registration_token) for _ in range(num)]
+        with Pool(processes=min(num, 10)) as pool:
+            instance_ids = pool.map(func=RunnerManager._create_runner, iterable=create_runner_args)
```

Comment:
Just out of curiosity, the exception handling here should be done by the RunnerManager static method? Then what happens to the instance_ids? should we filter out any that may have failed before running the tuple?

Reply:
I am testing out handling the exceptions for multiprocessing.

Reply:
This should be resolved with the new code using `imp_unordered`